theano/tensor/nnet/abstract_conv.py

```diff
@@ -1592,6 +1586,47 @@ def conv(self, img, kern, mode="valid", dilation=1, unshared=False):
             raise NotImplementedError('only 2D and 3D convolution are implemented')
         return out
 
+    def unshared2d(self, inp, kern, out_shape, direction=0):
+        '''
+        Basic slow Python unshared 2d convolution.
+        direction can be 0, 1 or 2 for forward pass,
+        gradWeights or gradInputs respectively
+        '''
+        if self.convdim != 2:
+            raise NotImplementedError('Unshared convolution not implemented for %dD'
+                                      % self.convdim)
+        if direction == 0:
+            if (kern.shape[0], kern.shape[1]) != (out_shape[0], out_shape[1]):
+                raise ValueError('Kernel shape ({},{}) does not match '
+                                 'output size ({},{})'.format(kern.shape[0], kern.shape[1],
+                                                              out_shape[0], out_shape[1]))
+        if direction not in (0, 1, 2):
+            raise ValueError("Direction must be one of 0, 1 or 2.")
+
+        out = np.zeros(out_shape, dtype=inp.dtype)
+
+        for row in xrange(out_shape[0]):
+            for col in xrange(out_shape[1]):
+                if direction == 0:
+                    out[row, col] += np.sum(np.multiply(inp[row:row + kern.shape[2],
+                                                        col:col + kern.shape[3]],
+                                            kern[row, col, ::-1, ::-1]))
+                elif direction == 1:
+                    out[row, col, ...] += kern[row, col] * \
+                        inp[row:row + out_shape[2], col:col + out_shape[3]]
+
+                else:  # direction = 2
+                    for k_row in xrange(kern.shape[2]):
+                        for k_col in xrange(kern.shape[3]):
+                            # Deciding which kernel to select
+                            reg_row = row - ((kern.shape[2] - 1) - k_row)
+                            reg_col = col - ((kern.shape[3] - 1) - k_col)
+                            if (reg_row >= 0 and reg_row < kern.shape[0]) and \
+                                    (reg_col >= 0 and reg_col < kern.shape[1]):
+                                out[row, col] += kern[reg_row, reg_col, k_row, k_col] * \
+                                    inp[reg_row, reg_col]
```

Comment:
This certainly looks a lot simpler!

But for the common `unshared2d` function, the outer two loops are common for all three directions (they go from `0` to `out_shape[0]` and `0` to `out_shape[1]`). We need them to go from `0` to `kern.shape[0]` and `0` to `kern.shape[1]`.

I think the easiest way to fix this is to have a condition to check whether `row` and `col` are less than `kern.shape[0]` and `kern.shape[1]` respectively, rather than modifying the loops for `k_row` and `k_col`.

The conditions will look much simpler and cleaner also.